### ISSUE TITLE
Hundreds of sockets in CLOSE_WAIT status

### ISSUE THREAD

**Issue by chris-david-taylor** [elastic/logstash#4267]
Hi All,

I am having a problem with an full ELK stack stopping accepting logs;  around 12-18 hours after starting logstash, the process will stop accepting logs and we'll typically see 500-600 sockets in CLOSE_WAIT status.  The main process doesn't appear to die.  Often when finally gives up, is recieving a very small amount of logs; typically a few hundred events in total between 20 or so servers. 

We have the following setup:

Logserver:
=================
logstash-1.5.1-1
elasticsearch-1.6.0-1
kibana-4.2-1
java-1.7.0-openjdk
Redhat Enterprise Linux 6.7

Server spec: 8GB, 4 x 2.7GHz cores.  Logstash runs with a 4GB heap and 8 worker threads.  Thread dumps and config can be supplied if needed.

Often when the server hits a maximum with the sockets, the load average is quite low, around 0.7 - it can reach around 2 when we restart the logstash service and its catching up.  For the case of testing, we have removed all filters from the pipeline; we did have the multiline filter in use to capture JVM stack traces, this has been removed to stop the pipeline becomeing single threaded. Presently our usage and config are too simple for it to be a case of an underspecced VM and we seem to be at the conclusion that might be a bug in logstash itself. 

Can anyone reproduce this confirm if this is a known bug?

Thanks,
Chris.

**Comment by chris-david-taylor** [comment]
Update; we just saw this in a stacktrace but it seems suspect as neither the heap or memory itself is maxed out:
 
 There is insufficient memory for the Java Runtime Environment to continue.
 Cannot create GC thread. Out of system resources.
 An error report file with more information is saved as:
 /tmp/jvm-9551/hs_error.log

Also, there are around 900 threads just waiting.

**Comment by chris-david-taylor** [comment]
Hi,

Does anyone have suggestions on the way forward on this?

Thanks,
Chris.

**Comment by jsvd** [comment]
can you provide the config? there are known leakages that have been fixed in the past, such as  https://github.com/logstash-plugins/logstash-output-elasticsearch/issues/306

**Comment by chris-david-taylor** [comment]
I'm afraid we had to migrate to Splunk before Christmas because my client couldn't wait any longer and called time, so I'm afraid the VM and config are gone (commercial realities).

Hopefully this will assist:   What I can recall is that we at one point we had the multi-line codec in use - removing that did improve the matter slightly.  We did try going back to basics with a clean VM and 'out of the box' install, with no configuration changes, but the problem persisted.  

The only suggestion that I can probably think of is that we don't have the worlds greatest SAN, and the rate of iops might be a little lower than we'd like, although we would think that having to monitor less than 20 VMs with very little logging (about 100 events per minute), would be reasonable.  I don't recall the record structure used by Splunk, but it does appear be keeping up with the read/writes it needs to perform.  If I find anything in our git repo that can help, I'll post it, but I'm afraid I've no longer got an environment for testing suggested changes/updated versions.

Thanks,
Chris.
